[GPIO] Improve pca_series gpio expander driver and retire duplicated drivers

CkovMk opened the pull request:
- Improves device reset
- Adds support for pca6408, pca6416, pcal9538, pca9539, pcal6408, pcal6416, alignd with Linux DT compatible strings.
- Retires pca6416 driver.
- pcal6408a and pcal6416a are already supported by "gpio_pcal64xxa" driver, which will coexist with this driver. Because pcal6408 and pcal6416 are registered-compatible with pcal9538 and pcal9539, and thete's no additional overhead to support them in this "gpio_pca_series" driver. The ultimate goal would be to deprecate "gpio_pcal64xxa" driver as well, but I don't have the resources to test the boards (mainly from Nordic) using it.
- Adds device specific flags extension to configure drive strength on supported devices (currently pcal6524 and pcal6534)

CkovMk committed:
drivers: gpio: pca_series: select i2c when enabled

On certain platforms I2C is not enabled by default, causing build error.
Change pca_series driver to select I2C.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk committed:
drivers: gpio: pca_series: add drive_strength config

Enables pca_series driver to set drive strength for part no with
PCA_HAS_LATCH capability.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

JiafeiPan commented on dts/bindings/gpio/nxp,pca6416a.yaml:
```diff
@@ -0,0 +1,10 @@
+# Copyright 2024-2025 NXP
+# SPDX-License-Identifier: Apache-2.0
+description: NXP PCA6416A I2C GPIO expander node
+compatible: nxp,pca6416a
```
one general comments, it will be better to not change dts compatible string, so that it will not break exiting dts nodes building (for example some offtrain application code).

CkovMk replied:
It aligns with the convension that pca(l)64xxa devices adds the "a" suffix. Either change the pca6416 to pca6416a or change both pcal6408a and pcal6416a to pcal6408 and pcal6416. I'd like to make it consistent.

CkovMk replied:
And the pcal64xxa driver has a much longer history.

CkovMk replied:
Alternatively, we can align the compatible string with Linux:

pca6408 (reuesd)
pcal6408/6416 (new, coexist with existing pcal6408a/6416a)

This also solves the problem of deprecating pcal64xxa driver

henrikbrixandersen replied:
I would prefer that.

JiafeiPan reviewed (commented):
Maybe could define some configure item for each chip if we use single driver to cover multiple chips, so that we could use these configure items to include some code in the driver in order to reduce the code size, it is helpful especially for some resource limited device.

CkovMk commented:
> Maybe could define some configure item for each chip if we use single driver to cover multiple chips, so that we could use these configure items to include some code in the driver in order to reduce the code size, it is helpful especially for some resource limited device.

Sure, it is planned. but I believe it's beyond the scope of this pr.
By comparing against the pcal64xxa driver, it merely adds RAM usage even when full register cache is enabled, and adds ~0.6KB ROM compared to pcal64xxa driver. What's more, only ~0.5KB ROM is required when a new device is supported, compared with the cost of 5.8KB if adding another driver.

Here's the result I get when supporting 1x PCAL6524 and 1x PCAL6408A on FRDM-IMX93 A55 board.

|  | Original | pca_series driver |
|---|---|---|
| RAM | 146B + 179B | 327B |
| ROM | 5.84KB + 6.47KB | 7.00KB  |

* In original state, it costs 146B + 5.84KB for pcal64xxa driver, 179B + 6.47KB for pca_series driver.

CkovMk committed:
drivers: gpio: pca_series: add support for pcal953x and pcal64xx

add support for pca9538, pcal9539.

Add support for pcal6408 and pcal6416, which is originally supported
by pcal64xxa driver. These device has the same register layout as
pcal9538 and pcal9539 respectively, which means they can be seamlessly
supported by pca_series driver.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk committed:
drivers: gpio: retire pca6416a driver

pca6416a driver can be replaced by pca_series driver, which covers a
larger number of devices.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk commented:
Sync compatible with Linux and rebase to main. This allows coexistence with pcal64xxa driver.

CkovMk commented:
Further split reset function

CkovMk committed:
drivers: gpio: pca_series: imporve reset function

The reset function is originally designed to have no return value. If
the GPIO toggle fails, there's no indication for failure until I2C
transaction fails. Account for this by adding return value to the reset
function.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk commented:
Tested PCA6408 input/output on FRDM-IMX93 and I.MX8MP EVK

CkovMk commented:
re-trigger ci

CkovMk commented:
updated title and description to better describe the purpose

pdgendt reviewed (approved):
Nit: `s/imporve/improve/` in third commit subject
